### ISSUE TITLE
[Talkback] Input decoration text is announced after text content

### ISSUE THREAD

**Issue by anu70** [flutter/flutter#110158]
When using Talkback for TextField, the prefix text and hint text are read after the input text (text entered by the user) is read. As a result, the user feels that the text is read backward.

Sample flutter app :
[demo app talkabck.webm](https://user-images.githubusercontent.com/15100500/186377081-59ad3e17-f7b7-41ad-9959-91c3d12622f1.webm)

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/cupertino.dart';
import 'package:mobile.flutter.google_material.material2/google_material.dart';
 
void main() => runApp(MyApp());
 
class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: WelcomScreenWidget()
    );
  }
}
 
class WelcomScreenWidget extends StatefulWidget {
  const WelcomScreenWidget({Key? key}) : super(key: key);
 
  @override
  State<WelcomScreenWidget> createState() => _WelcomScreenWidgetState();
}
 
class _WelcomScreenWidgetState extends State<WelcomScreenWidget> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.grey,
      appBar: AppBar(
        title: Text("Welcome"),
      ),
      body: _buildCardNumberInputField(context, TextEditingController()),
    );
  }
 
  Widget _buildCardNumberInputField(
      BuildContext context,
      TextEditingController controller,
      ) {
    final textTheme = GoogleTextTheme.of(context);
    final cardNumberTextTheme =
    textTheme.gmBody1!;
 
    return Column(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        TextField(
          textAlign: TextAlign.left,
          style: cardNumberTextTheme,
          maxLines: 1,
          decoration: InputDecoration(
            fillColor: Colors.transparent,
            enabledBorder:  OutlineInputBorder(
              borderSide: BorderSide(color: Theme.of(context).dividerColor),
            ),
            focusedBorder:  OutlineInputBorder(
              borderSide: BorderSide(color: Theme.of(context).primaryColor, width: 2),
            ),
            labelText: "label text",
            labelStyle: cardNumberTextTheme,
            floatingLabelBehavior: FloatingLabelBehavior.never,
            hintText: "hint text",
            hintStyle: cardNumberTextTheme,
            prefixText: "prefix text",
            prefixStyle: cardNumberTextTheme,
          ),
          controller: controller,
          keyboardType: TextInputType.number,
        ),
    ]
    );
  }
} 
```
</details>

**Comment by danagbemava-nc** [comment]
I can reproduce the behaviour of the labels being announced after the text on android. On iOS, the labels/prefix are read before the text content is announced.

Labeling for further insight from the team. 

<details>
<summary>video</summary>

| android | iOS | 
| -- | -- |
| <video src="https://user-images.githubusercontent.com/88313112/186429395-16552ae5-e440-4722-94a4-5a46ffc396b8.mp4" /> | <video src="https://user-images.githubusercontent.com/88313112/186429428-170488e9-0039-404c-b5c5-4a55228b35e1.MP4" /> |


</details>

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MaterialApp(home: MyApp()));

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Plugin example app'),
      ),
      body: TextField(
        autofocus: true,
        controller: TextEditingController(text: "some text"),
        decoration: const InputDecoration(
          prefixText: "prefix",
          labelText: "Label",
        ),
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.5 21G72 darwin-arm, locale en-GB)
    • Flutter version 3.0.5 at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (6 weeks ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.46.0

[✓] Connected device (3 available)
    • M2007J20CG (mobile) • 5dd3be00 • android-arm64  • Android 11 (API 30)
    • macOS (desktop)     • macos    • darwin-arm64   • macOS 12.5 21G72 darwin-arm
    • Chrome (web)        • chrome   • web-javascript • Google Chrome 104.0.5112.101

[!] HTTP Host Availability
    ✗ HTTP host "https://cocoapods.org/" is not reachable. Reason: An error occurred while checking the HTTP host: Connection reset by peer

! Doctor found issues in 1 category.
```

```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.2443, on macOS 12.5 21G72 darwin-arm, locale en-GB)
    • Flutter version 3.1.0-0.0.pre.2443 on channel master at /Users/nexus/dev/sdks/flutters
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e45031bf2a (7 hours ago), 2022-08-24 01:59:20 -0400
    • Engine revision 55c461f286
    • Dart version 2.19.0 (build 2.19.0-130.0.dev)
    • DevTools version 2.16.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13F100
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.46.0

[✓] Connected device (3 available)
    • M2007J20CG (mobile) • 5dd3be00 • android-arm64  • Android 11 (API 30)
    • macOS (desktop)     • macos    • darwin-arm64   • macOS 12.5 21G72 darwin-arm
    • Chrome (web)        • chrome   • web-javascript • Google Chrome 104.0.5112.101

[!] HTTP Host Availability
    ✗ HTTP host "https://cocoapods.org/" is not reachable. Reason: An error occurred while checking the HTTP host: Connection reset by peer

! Doctor found issues in 1 category.
```
</details>

**Comment by goderbauer** [comment]
cc @justinmc @chunhtai

**Comment by goderbauer** [comment]
Internal issue: b/240919433

**Comment by justinmc** [comment]
@chunhtai Should we add in the missing "editing" word as a part of this issue or should we consider this working as intended?

Also @TahaTesser thanks for the excellent investigation here!

**Comment by TahaTesser** [comment]
There is another slight difference, the Native says  "editing text_content edit box label" and Flutter says "text_content label edit box" if that's something to consider too.

**Comment by chunhtai** [comment]
I traced this back to how we concate label value and hint in accessibility bridge that were introduced 5 years ago, the plan is to match iOS feature, but there isn't much explanation about the ordering. @goderbauer do you remember why value is ordered ahead of label?

@TahaTesser i bet there is an additional field in accessibilitynodeinfo we forgot to set. I have to looking talkback source code to be sure. can you file a separate issue for it?

**Comment by chunhtai** [comment]
actually as the @TahaTesser pointed out the editing keyword may be the only issue here. I will investigate a bit more

**Comment by chunhtai** [comment]
finished investigation. It looks like the prefix and suffix have their own accessibility node in native Android.

This is the code i am using
```xml
    <com.google.android.material.textfield.TextInputLayout
        android:id="@+id/textField"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:hint="label"
        app:prefixText="prefix"
        app:suffixText="sufix">
        <com.google.android.material.textfield.TextInputEditText
            android:layout_width="match_parent"
            android:layout_height="wrap_content"
            android:text="text" />

    </com.google.android.material.textfield.TextInputLayout>
```

The focus order in talk back is
prefix (speak: <prefix> )-> textfield (speak: <text> edit box <label>, double tap to edit text) -> suffix (speak: <suffix>)


To match this behavior, the input decoration needs to create separate semantics node for prefix and suffix.

**Comment by chunhtai** [comment]
iOS has similar behavior, This is the code I am using
```objc
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    UITextField *textField = [[UITextField alloc] initWithFrame:CGRectMake(10, 200, 300, 40)];
    textField.borderStyle = UITextBorderStyleRoundedRect;
    textField.font = [UIFont systemFontOfSize:15];
    textField.placeholder = @"enter text";
    textField.autocorrectionType = UITextAutocorrectionTypeNo;
    textField.keyboardType = UIKeyboardTypeDefault;
    textField.returnKeyType = UIReturnKeyDone;
    textField.clearButtonMode = UITextFieldViewModeWhileEditing;
    textField.contentVerticalAlignment = UIControlContentVerticalAlignmentCenter;
    textField.text = @"some text";
    UILabel* prefix = [[UILabel alloc] init];
    prefix.text = @"prefix";
    // set font, color etc.
    [prefix sizeToFit];
    UILabel* suffix = [[UILabel alloc] init];
    suffix.text = @"suffix";
    // set font, color etc.
    [suffix sizeToFit];
    textField.leftView = prefix;
    textField.leftViewMode = UITextFieldViewModeAlways;
    textField.rightView = suffix;
    textField.rightViewMode = UITextFieldViewModeAlways;
    [self.view addSubview:textField];
}
```

**Comment by chunhtai** [comment]
update: still working on the pr

**Comment by chunhtai** [comment]
Update: There is another issue that may require changes to the merge rule https://github.com/flutter/flutter/issues/111931 I am currently trying to figure out an API that can solve both issues.

**Comment by chunhtai** [comment]
update: still working on the PR

**Comment by GaryQian** [comment]
(From Android triage) @chunhtai Any progress with the PR?

**Comment by chunhtai** [comment]
https://github.com/flutter/flutter/pull/110730 is in review.
